GitCommands/PathUtil.cs

```diff
@@ -124,14 +124,42 @@ public static string NormalizePath([NotNull] this string path)
 
             try
             {
-                return Path.GetFullPath(new Uri(path).LocalPath);
+                return Path.GetFullPath(Resolve(path));
             }
             catch (UriFormatException)
             {
                 return string.Empty;
             }
         }
 
+        [NotNull]
+        public static string Resolve([NotNull] string path, string relativePath = "")
+        {
+            if (string.IsNullOrWhiteSpace(path))
+            {
+                throw new ArgumentException(nameof(path));
+            }
+
+            // Fix for on purpose invalid machine name of WSL in Windows 10
+            bool replaceWslPath = path.StartsWith(@"\\wsl$\");
+            if (replaceWslPath)
+            {
+                // Temporarily replace machine name with a valid name
+                path = Regex.Replace(path, "^" + Regex.Escape(@"\\wsl$\"), @"\\wsl\");
+            }
+
+            string localPath =
+                (string.IsNullOrEmpty(relativePath) ? new Uri(path) : new Uri(new Uri(path), relativePath)).LocalPath;
+
+            if (replaceWslPath)
+            {
+                // Revert temporary replacement of WSL machine name
+                localPath = Regex.Replace(localPath, "^" + Regex.Escape(@"\\wsl\"), @"\\wsl$\");
+            }
```

Comment:
A number of unnecessary allocations
```suggestion
			Uri tempPath = new Uri(path);
			if (string.IsNullOrEmpty(relativePath)
			{
				tempPath = new Uri(tempPath, relativePath);
			}
			
            if (!replaceWslPath)
            {
            	return tempPath.LocalPath;
            }

            // Revert temporary replacement of WSL machine name
            return tempPath.LocalPath.Replace(@"\\wsl\", @"\\wsl$\");
```